yarn-project/stdlib/src/block/l2_block_downloader/l2_block_stream.ts

```diff
@@ -134,6 +144,15 @@ export class L2BlockStream {
       return true;
     }
     const localBlockHash = await this.localData.getL2BlockHash(blockNumber);
+    if (!localBlockHash && this.opts.skipFinalized) {
+      // Failing to find a block hash when skipping finalized blocks can be highly problematic as we'd potentially need
+      // to go all the way back to the genesis block to find a block in which we agree with the source (since we've
+      // potentially skipped all history). This means that stores that prune old blocks must be careful to leave no gaps
+      // when going back from latest block to the last finalized one.
+      this.log.error(`No local block hash for block number ${blockNumber}`);
+      throw new AbortError();
+    }
```

Comment:
I went back and forth on this. Ultimately I think I'd rather fail loudly if we run into this scenario instead of spending multiple hours going back in time one block at a time until we find a block in which we agree with the source.

Reply:
This will break if a reorg hits and prunes the chain to before the `PXE_L2_STARTING_BLOCK`, since there weill be no local data for that block number (fun fact: this is similar to the issue I fixed in #13473, but for local instead of source). 

I don't know if we have any setup where the starting block is not 1, but if there is, it will trigger this exception.

Reply:
I'm thinking that with this new option we could very well get rid of this envvar. Message discovery is unaffected by the starting block since we search tagged logs as opposed to scanning all blocks, and once this is merged we will no longer process all blocks in any flow.

Reply:
Opened https://github.com/AztecProtocol/aztec-packages/pull/13504